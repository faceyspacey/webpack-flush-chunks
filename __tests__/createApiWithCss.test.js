// @noflow
import {
  default as createApiWithCss,
  getJsFileRegex,
  isJs,
  isCss,
  stylesAsString
} from '../src/createApiWithCss'

jest.mock('fs', () => ({
  readFileSync: fileName =>
    `${fileName}- the css! \n/*# sourceMappingURL=/static/main.js.map */`
}))

const publicPath = '/static/'
const outputPath = '/Users/jamesgillmore/App/build'

/** CREATE API WITH CSS */

describe('createApiWithCss()', () => {
  it('generates js + style components, strings and arrays', () => {
    const files = ['0.js', '0.css', 'main.js', 'main.css']
    const filesForCss = ['main.js', 'main.css', '0.js', '0.css']
    const api = createApiWithCss(
      files,
      filesForCss,
      publicPath,
      outputPath
    ) /*? $ */

    expect(api.Js() /*? $.props.children */).toMatchSnapshot()
    expect(api.Styles() /*? $.props.children */).toMatchSnapshot()
    expect(api.Css() /*? $.props */).toMatchSnapshot()

    expect(`${api.js}` /*? */).toMatchSnapshot()
    expect(`${api.styles}` /*? */).toMatchSnapshot()
    expect(`${api.css}` /*? */).toMatchSnapshot()

    expect(api.scripts /*? $ */).toMatchSnapshot()
    expect(api.stylesheets /*? $ */).toMatchSnapshot()

    expect(api.publicPath).toEqual(publicPath.replace(/\/$/, ''))
    expect(api.outputPath).toEqual(outputPath)
  })

  it('uses files with extension "no_css.js" if available', () => {
    const files = ['main.js', 'main.no_css.js', 'main.css']
    const api = createApiWithCss(files, files, publicPath, outputPath) /*? $ */

    expect(api.Js() /*? $.props.children */).toMatchSnapshot()
    expect(api.Styles() /*? $.props.children */).toMatchSnapshot()
    expect(api.Css() /*? $.props */).toMatchSnapshot()

    expect(`${api.js}` /*? */).toMatchSnapshot()
    expect(`${api.styles}` /*? */).toMatchSnapshot()
    expect(`${api.css}` /*? */).toMatchSnapshot()

    expect(api.scripts /*? $ */).toMatchSnapshot()
    expect(api.stylesheets /*? $ */).toMatchSnapshot()

    expect(api.publicPath).toEqual(publicPath.replace(/\/$/, ''))
    expect(api.outputPath).toEqual(outputPath)
  })

  it('throws when rendering css without outputPath', () => {
    const files = ['main.js', 'main.css']
    const api = createApiWithCss(files, files, publicPath)

    expect(api.Css /*? */).toThrow()
    expect(api.css.toString /*? */).toThrow()
  })

  it('adds trailing slash to public path', () => {
    const files = ['main.js']
    const publicPath = '/static'
    const api = createApiWithCss(files, files, publicPath)

    expect(api.js.toString() /*? */).toContain('/static/main.js')
  })

  it('does not include scripts with extension "hot-update.js"', () => {
    const files = ['main.js', 'main.hot-update.js']
    const api = createApiWithCss(files, files, publicPath)

    expect(api.scripts /*? */).not.toContain('main.hot-update.js')
  })
})

/** HELPERS */

describe('unit tests', () => {
  test('getJsFileRegex()', () => {
    let files = ['main.no_css.js', 'main.js', 'main.css']
    let regex = getJsFileRegex(files)
    expect(regex).toEqual(/\.no_css\.js$/)

    files = ['main.js', 'main.css']
    regex = getJsFileRegex(files)
    expect(regex).toEqual(/\.js$/)
  })

  test('isJs()', () => {
    const jsRegex = /\.js$/
    const noCssRegex = /\.no_css\.js$/

    expect(isJs(jsRegex, 'main.js')).toEqual(true)
    expect(isJs(noCssRegex, 'main.js')).toEqual(false)

    expect(isJs(noCssRegex, 'main.no_css.js')).toEqual(true)
    expect(isJs(noCssRegex, 'main.js')).toEqual(false)

    expect(isJs(jsRegex, 'main.hot-update.js')).toEqual(false)
    expect(isJs(noCssRegex, 'main.hot-update.js')).toEqual(false)
  })

  test('isCss()', () => {
    expect(isCss('main.css')).toEqual(true)
    expect(isCss('main.js')).toEqual(false)
  })

  test('stylesAsString()', () => {
    const stylesheets = ['main.css', '0.css']
    const css = stylesAsString(stylesheets, outputPath) /*? $ */
    expect(css).toMatchSnapshot()
  })
})
