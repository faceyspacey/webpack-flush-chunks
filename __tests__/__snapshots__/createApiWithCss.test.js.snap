// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createApiWithCss() generates js + style components, strings and arrays 1`] = `
<span>
  <script
    src="/static/0.js"
    type="text/javascript"
  />
  <script
    src="/static/main.js"
    type="text/javascript"
  />
</span>
`;

exports[`createApiWithCss() generates js + style components, strings and arrays 2`] = `
<span>
  <link
    href="/static/main.css"
    rel="stylesheet"
  />
  <link
    href="/static/0.css"
    rel="stylesheet"
  />
</span>
`;

exports[`createApiWithCss() generates js + style components, strings and arrays 3`] = `
<span>
  <style>
    /Users/jamesgillmore/App/build/main.css- the css! 
    
    /Users/jamesgillmore/App/build/0.css- the css! 
    
  </style>
</span>
`;

exports[`createApiWithCss() generates js + style components, strings and arrays 4`] = `
"<script type='text/javascript' src='/static/0.js'></script>
<script type='text/javascript' src='/static/main.js'></script>"
`;

exports[`createApiWithCss() generates js + style components, strings and arrays 5`] = `
"<link rel='stylesheet' href='/static/main.css' />
<link rel='stylesheet' href='/static/0.css' />"
`;

exports[`createApiWithCss() generates js + style components, strings and arrays 6`] = `
"<style>/Users/jamesgillmore/App/build/main.css- the css! 

/Users/jamesgillmore/App/build/0.css- the css! 
</style>"
`;

exports[`createApiWithCss() generates js + style components, strings and arrays 7`] = `
Array [
  "0.js",
  "main.js",
]
`;

exports[`createApiWithCss() generates js + style components, strings and arrays 8`] = `
Array [
  "main.css",
  "0.css",
]
`;

exports[`createApiWithCss() uses files with extension "no_css.js" if available 1`] = `
<span>
  <script
    src="/static/main.no_css.js"
    type="text/javascript"
  />
</span>
`;

exports[`createApiWithCss() uses files with extension "no_css.js" if available 2`] = `
<span>
  <link
    href="/static/main.css"
    rel="stylesheet"
  />
</span>
`;

exports[`createApiWithCss() uses files with extension "no_css.js" if available 3`] = `
<span>
  <style>
    /Users/jamesgillmore/App/build/main.css- the css! 
    
  </style>
</span>
`;

exports[`createApiWithCss() uses files with extension "no_css.js" if available 4`] = `"<script type='text/javascript' src='/static/main.no_css.js'></script>"`;

exports[`createApiWithCss() uses files with extension "no_css.js" if available 5`] = `"<link rel='stylesheet' href='/static/main.css' />"`;

exports[`createApiWithCss() uses files with extension "no_css.js" if available 6`] = `
"<style>/Users/jamesgillmore/App/build/main.css- the css! 
</style>"
`;

exports[`createApiWithCss() uses files with extension "no_css.js" if available 7`] = `
Array [
  "main.no_css.js",
]
`;

exports[`createApiWithCss() uses files with extension "no_css.js" if available 8`] = `
Array [
  "main.css",
]
`;

exports[`unit tests stylesAsString() 1`] = `
"/Users/jamesgillmore/App/build/main.css- the css! 

/Users/jamesgillmore/App/build/0.css- the css! 
"
`;
